kratos/processes/fast_transfer_between_model_parts_process.cpp

```diff
@@ -347,80 +338,84 @@ void FastTransferBetweenModelPartsProcess::ReplicateWithFlags()
+                auto it_elem = it_elem_begin + i;
                 if (it_elem->Is(mFlag)) {
                     Element::Pointer p_new_elem = it_elem->Clone(total_num_elements + i + 1, it_elem->GetGeometry());
-                    (elements_buffer_vector[thread_id]).insert((elements_buffer_vector[thread_id]).begin(), p_new_elem);
+                    (elements_buffer_vector).insert(elements_buffer_vector.begin(), p_new_elem);
                 }
             }
         }
 
         if (num_conditions != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONDITIONS || mEntity == EntityTransfered::NODESANDCONDITIONS)) {
+            const auto it_cond_begin = mrOriginModelPart.ConditionsBegin();
             #pragma omp for
             for(int i = 0; i < num_conditions; ++i) {
-                auto it_cond = mrOriginModelPart.ConditionsBegin() + i;
+                auto it_cond = it_cond_begin + i;
                 if (it_cond->Is(mFlag)) {
                     Condition::Pointer p_new_cond = it_cond->Clone(total_num_conditions + i + 1, it_cond->GetGeometry());
-                    (conditions_buffer_vector[thread_id]).insert((conditions_buffer_vector[thread_id]).begin(), p_new_cond);
+                    (conditions_buffer_vector).insert(conditions_buffer_vector.begin(), p_new_cond);
                 }
             }
         }
 
         if (num_constraints != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::CONSTRAINTS || mEntity == EntityTransfered::NODESANDCONSTRAINTS)) {
+            const auto it_const_begin = mrOriginModelPart.MasterSlaveConstraintsBegin();
             #pragma omp for
             for(int i = 0; i < num_constraints; ++i) {
-                auto it_const = mrOriginModelPart.MasterSlaveConstraintsBegin() + i;
+                auto it_const = it_const_begin + i;
                 if (it_const->Is(mFlag)) {
                     MasterSlaveConstraint::Pointer p_new_const = it_const->Clone(total_num_constraints + i + 1);
-                    (constraints_buffer_vector[thread_id]).insert((constraints_buffer_vector[thread_id]).begin(), p_new_const);
+                    (constraints_buffer_vector).insert(constraints_buffer_vector.begin(), p_new_const);
                 }
             }
         }
 
         // We add to the model part
-        #pragma omp single
+        #pragma omp critical
         {
             if (num_nodes != 0 && (mEntity == EntityTransfered::ALL || mEntity == EntityTransfered::NODES || mEntity == EntityTransfered::NODESANDELEMENTS || mEntity == EntityTransfered::NODESANDCONDITIONS))
-                for( auto& node_buffer_vector : nodes_buffer_vector)
-                    mrDestinationModelPart.AddNodes(node_buffer_vector.begin(),node_buffer_vector.end());
+                mrDestinationModelPart.AddNodes(nodes_buffer_vector.begin(),nodes_buffer_vector.end());
         }
 
-        #pragma omp single
+        #pragma omp critical
```

Comment:
one single critical section. to make it clear i think it would make sense to have multiple "single" pragmas but if you put critical than my impression is that a single one is better

Reply:
@roigcarlo can you clarify?

Reply:
After speaking with @roigcarlo in private I changed and unified the criticals